syft/execution/role.py

```diff
@@ -312,6 +313,100 @@ def copy(self):
             id=sy.ID_PROVIDER.pop(),
         )
 
+    def _is_inplace_operation(self, cmd: str, target: PlaceholderId):
+        """
+        Helper method that returns True if cmd is inplace operation
+        """
+        if target is None:
+            framework_name, command = cmd.split(".", 2)
+            framework = getattr(syft, framework_name, syft.framework)
+            return framework.is_inplace_method(command)
+
+        elif isinstance(target, PlaceholderId):
+            ph = self.placeholders.get(target.value, None)
+            if ph is not None:
+                framework_name = ph.child.__module__
+                framework = getattr(syft, framework_name, syft.framework)
+                return framework.is_inplace_method(cmd)
+
+        return False
+
+    def _prune_actions(self):
+        """
+        Removes unnecessary actions and placeholders.
+        """
+
+        # Track operations on output placeholders from bottom to top
+        output_connected_placeholder_ids = set(self.output_placeholder_ids)
+        connected_actions_idx = set()
+        for action_rev_idx, action in enumerate(reversed(self.actions)):
+            return_ids = (
+                {ph.value for ph in action.return_ids if isinstance(ph, PlaceholderId)}
+                if action.return_ids is not None
+                else set()
+            )
```

Comment:
This is very nested and is not optimal for readability
```suggestion
            return_ids = set()
            if action.return_ids is not None:
                for ph in action.return_ids:
                    if isinstance(ph, PlaceholderId):
                        return_ids.add(ph.value)
```